File: src/main/java/main/commands/commands/util/SuggestionCommand.java
package main.commands.commands.util;

import lombok.SneakyThrows;
import main.commands.CommandContext;
import main.commands.ICommand;
import main.main.Robertify;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.component.InteractiveCommand;
import main.utils.database.mongodb.cache.BotInfoCache;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
import net.dv8tion.jda.api.exceptions.ErrorHandler;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.requests.ErrorResponse;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.script.ScriptException;
import java.awt.*;
import java.time.Instant;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class SuggestionCommand extends InteractiveCommand implements ICommand {
    private final static Logger logger = LoggerFactory.getLogger(SuggestionCommand.class);

    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final List<String> args = ctx.getArgs();
        final Message msg = ctx.getMessage();

        if (args.isEmpty()) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(msg.getGuild(), "You must provide a suggestion!").build())
                    .queue();
        } else {
            switch (args.get(0).toLowerCase()) {
                case "setup" -> setup(msg);
                case "accept" -> accept(msg, args);
                case "deny", "reject" -> deny(msg, args);
                case "ban" -> ban(msg, args);
                case "unban" -> unban(msg, args);
                default -> sendSuggestion(msg, args);
            }
        }
    }

    private void setup(Message msg) {
        if (!isDeveloper(msg.getAuthor())) return;

        final var config = BotInfoCache.getInstance();

        if (config.isSuggestionsSetup()) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(msg.getGuild(), "The suggestions channels have already been setup!").build())
                    .queue();
            return;
        }

        final Guild guild = msg.getGuild();

        guild.createCategory("Suggestions").queue(category -> {

           category.upsertPermissionOverride(guild.getPublicRole())
                   .setDeny(Permission.VIEW_CHANNEL)
                   .queueAfter(1, TimeUnit.SECONDS);

           guild.createTextChannel("pending-suggestions", category)
                   .queueAfter(2, TimeUnit.SECONDS, pendingChannel -> {
                       guild.createTextChannel("accepted-suggestions", category)
                               .queueAfter(1, TimeUnit.SECONDS, acceptedChannel -> {
                                  guild.createTextChannel("denied-suggestions", category)
                                          .queueAfter(1, TimeUnit.SECONDS, deniedChannel -> {
                                              config.initSuggestionChannels(category.getIdLong(), pendingChannel.getIdLong(), acceptedChannel.getIdLong(), deniedChannel.getIdLong());
                                                msg.addReaction("✅").queue();
                                          });
                               });
                   });
        });

    }

    private void sendSuggestion(Message msg, List<String> args) {
        msg.replyEmbeds(handleSuggestion(msg.getGuild(), msg.getAuthor(), String.join(" ", args)))
                .queue(sentMessage -> {
                    msg.delete().queueAfter(10, TimeUnit.SECONDS);
                    sentMessage.delete().queueAfter(11, TimeUnit.SECONDS);
                });
    }

    private void accept(Message msg, List<String> args) {
        if (!isDeveloper(msg.getAuthor())) return;

        final var guild = msg.getGuild();

        if (args.size() < 2) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must provide the ID of a suggestion to accept").build())
                    .queue();
            return;
        }

        String id = args.get(1);

        if (!GeneralUtils.stringIsID(id)) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "Invalid ID!").build())
                    .queue();
            return;
        }

        final var config = BotInfoCache.getInstance();

        TextChannel pendingChannel = Robertify.api.getTextChannelById(config.getSuggestionsPendingChannelID());

        if (pendingChannel == null) {
            logger.warn("The pending suggestions channel isn't setup!");
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "This feature isn't available right now!").build()).queue();
            return;
        }

        pendingChannel.retrieveMessageById(id).queue(suggestion -> {
            TextChannel acceptedChannel = Robertify.api.getTextChannelById(config.getSuggestionsAcceptedChannelID());

            if (acceptedChannel == null) {
                logger.warn("The pending suggestions channel isn't setup!");
                msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "This feature isn't available right now!").build()).queue();
                return;
            }
            List<MessageEmbed.Field> embedField = suggestion.getEmbeds().get(0).getFields();
            final String suggester = embedField.get(0).getValue();
            final String pendingSuggestion = embedField.get(1).getValue();

            acceptedChannel.sendMessageEmbeds(getEmbed(
                    new Color(77, 255, 69),
                    suggester,
                    pendingSuggestion,
                    suggestion.getEmbeds().get(0).getThumbnail().getUrl()
            )).queue();

            suggestion.delete().queue();

            GeneralUtils.retrieveUser(GeneralUtils.getDigitsOnly(suggester))
                    .openPrivateChannel().queue(channel -> {
                        final EmbedBuilder acceptedEmbed = new EmbedBuilder();
                        acceptedEmbed.setColor(new Color(77, 255, 69));
                        acceptedEmbed.setTitle("Suggestions");
                        acceptedEmbed.setDescription("**Your suggestion has been accepted!**" +
                                "\nYou will see it appear in the next changelog");
                        acceptedEmbed.addField("Suggestion", pendingSuggestion, false);

                        channel.sendMessageEmbeds(acceptedEmbed.build())
                                .queue(null, new ErrorHandler()
                                        .handle(ErrorResponse.CANNOT_SEND_TO_USER, ignored -> {}));
                    });

            msg.addReaction("✅").queue();
        }, new ErrorHandler()
                .handle(ErrorResponse.UNKNOWN_MESSAGE, e -> {
                    msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "That ID doesn't belong to any pending suggestion!").build())
                            .queue();
                }));
    }

    private void deny(Message msg, List<String> args) {
        if (!isDeveloper(msg.getAuthor())) return;

        final var guild = msg.getGuild();

        if (args.size() < 2) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must provide the ID of a suggestion to reject").build())
                    .queue();
            return;
        }

        final String id = args.get(1);
        final String reason;

        if (args.size() >= 3) reason = String.join(" ", args.subList(2, args.size()));
        else reason = null;

        if (!GeneralUtils.stringIsID(id)) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "Invalid ID!").build())
                    .queue();
            return;
        }

        final var config = BotInfoCache.getInstance();

        TextChannel pendingChannel = Robertify.api.getTextChannelById(config.getSuggestionsPendingChannelID());

        if (pendingChannel == null) {
            logger.warn("The pending suggestions channel isn't setup!");
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "This feature isn't available right now!").build()).queue();
            return;
        }

        pendingChannel.retrieveMessageById(id).queue(suggestion -> {
            TextChannel deniedChannel = Robertify.api.getTextChannelById(config.getSuggestionsDeniedChannelID());

            if (deniedChannel == null) {
                logger.warn("The pending suggestions channel isn't setup!");
                msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "This feature isn't available right now!").build()).queue();
                return;
            }
            List<MessageEmbed.Field> embedField = suggestion.getEmbeds().get(0).getFields();
            final String suggester = embedField.get(0).getValue();
            final String pendingSuggestion = embedField.get(1).getValue();

            deniedChannel.sendMessageEmbeds(getEmbed(
                    new Color(187, 0, 0),
                    suggester,
                    pendingSuggestion,
                    suggestion.getEmbeds().get(0).getThumbnail().getUrl()
            )).queue();

            suggestion.delete().queue();

            GeneralUtils.retrieveUser(GeneralUtils.getDigitsOnly(suggester))
                    .openPrivateChannel().queue(channel -> {
                        final EmbedBuilder deniedEmbed = new EmbedBuilder();
                        deniedEmbed.setTitle("Suggestions");
                        deniedEmbed.setColor(new Color(187, 0, 0));
                        deniedEmbed.setDescription("**Your suggestion has been denied**");
                        deniedEmbed.addField("Suggestion", pendingSuggestion, false);

                        if (reason != null)
                            deniedEmbed.addField("Reason", reason, false);

                        channel.sendMessageEmbeds(deniedEmbed.build())
                                .queue(null, new ErrorHandler()
                                        .handle(ErrorResponse.CANNOT_SEND_TO_USER, ignored -> {}));
                    });

            msg.addReaction("✅").queue();
        }, new ErrorHandler()
                .handle(ErrorResponse.UNKNOWN_MESSAGE, e -> {
                    msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "That ID doesn't belong to any pending suggestion!").build())
                            .queue();
                }));
    }

    private MessageEmbed getEmbed(Color color, String suggester, String suggestion, String thumbnail) {
        final EmbedBuilder eb = new EmbedBuilder();

        eb.setColor(color);
        eb.setTitle("Suggestion");
        eb.addField("Suggester",  suggester , false);
        eb.addField("Suggestion", suggestion, false);

        if (thumbnail != null)
            eb.setThumbnail(thumbnail);

        eb.setTimestamp(Instant.now());

        return eb.build();
    }

    private MessageEmbed handleSuggestion(Guild guild, User suggester, String suggestion) {
        if (suggestion.chars().count() > 1024)
            return RobertifyEmbedUtils.embedMessage(guild, "Your suggestion must be no more than 1024 characters!").build();

        final var config = BotInfoCache.getInstance();
        final TextChannel pendingChannel = Robertify.api.getTextChannelById(config.getSuggestionsPendingChannelID());

        if (pendingChannel == null) {
            logger.warn("The suggestion channels aren't setup!");
            return RobertifyEmbedUtils.embedMessage(guild, "This feature isn't available right now!").build();
        }

        if (config.userIsSuggestionBanned(suggester.getIdLong()))
            return RobertifyEmbedUtils.embedMessage(guild, "You have been banned from sending suggestions!").build();

        final EmbedBuilder eb = new EmbedBuilder();

        eb.setColor(new Color(255, 196, 0));
        eb.setTitle("Suggestion from " + suggester.getName());
        eb.addField("Suggester", suggester.getAsMention(), false);
        eb.addField("Suggestion", suggestion, false);
        eb.setThumbnail(suggester.getEffectiveAvatarUrl());
        eb.setTimestamp(Instant.now());

        pendingChannel.sendMessageEmbeds(eb.build()).queue();
        return RobertifyEmbedUtils.embedMessage(guild, "You have successfully sent your suggestion!\n\n" +
                "*Be sure to be on the lookout for a response in your DMs! Ensure your direct messages are enabled" +
                " else you will not receive any updates on your suggestion.*").build();
    }

    private void ban(Message msg, List<String> args) {
        if (!isDeveloper(msg.getAuthor())) return;

        final var guild = msg.getGuild();

        if (args.size() < 2) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must provide the ID of a user to ban").build())
                    .queue();
            return;
        }

        final String id = args.get(1);

        if (!GeneralUtils.stringIsID(id)) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "Invalid ID!").build())
                    .queue();
            return;
        }

        final var user = GeneralUtils.retrieveUser(GeneralUtils.getDigitsOnly(id));

        if (user == null) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "Invalid user!").build())
                    .queue();
            return;
        }

        try {
            BotInfoCache.getInstance().banSuggestionsUser(user.getIdLong());
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You have banned "
                    + user.getAsMention() + "from suggestions").build())
                    .queue();
        } catch (IllegalStateException e) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, e.getMessage()).build()).queue();
        } catch (Exception e) {
            logger.error("Unexpected error", e);
            msg.addReaction("❌").queue();
        }
    }

    private void unban(Message msg, List<String> args) {
        if (!isDeveloper(msg.getAuthor())) return;

        final var guild = msg.getGuild();

        if (args.size() < 2) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must provide the ID of a user to ban").build())
                    .queue();
            return;
        }

        final String id = args.get(1);

        if (!GeneralUtils.stringIsID(id)) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "Invalid ID!").build())
                    .queue();
            return;
        }

        final var user = GeneralUtils.retrieveUser(GeneralUtils.getDigitsOnly(id));

        if (user == null) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "Invalid user!").build())
                    .queue();
            return;
        }

        try {
            BotInfoCache.getInstance().unbanSuggestionUser(user.getIdLong());
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You have banned "
                            + user.getAsMention() + "from suggestions").build())
                    .queue();
        } catch (IllegalStateException e) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, e.getMessage()).build()).queue();
        } catch (Exception e) {
            logger.error("Unexpected error", e);
            msg.addReaction("❌").queue();
        }
    }

    @SneakyThrows
    private boolean isDeveloper(User user) {
        return BotInfoCache.getInstance().isDeveloper(user.getIdLong());
    }

    @Override
    public String getName() {
        return "suggest";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+GeneralUtils.listToString(getAliases())+"`\n" +
                "Want to see something added to Robertify? Suggest it using this command!\n\n" +
                "**Usage**: `"+ prefix +"suggest <suggestion>`";
    }

    @Override
    public List<String> getAliases() {
        return List.of("sg", "sug");
    }

    @Override
    public void initCommand() {

    }

    @Override
    public void initCommand(Guild g) {
        setInteractionCommand(getCommand());
        upsertCommand(g);
    }

    public InteractionCommand getCommand() {
        return InteractionCommand.create()
                .setCommand(Command.of(
                        getName(),
                        "Suggest a feature you'd like to see in the bot!",
                        List.of(CommandOption.of(
                                OptionType.STRING,
                                "suggestion",
                                "The suggestion to send to us",
                                true
                        ))
                )).build();
    }

    @Override
    public void onSlashCommand(@NotNull SlashCommandEvent event) {
        if (!event.getName().equals(getName())) return;

        event.replyEmbeds(handleSuggestion(event.getGuild(), event.getUser(), event.getOption("suggestion").getAsString()))
                .setEphemeral(true)
                .queue();
    }
}

File: src/main/kotlin/main/commands/slashcommands/util/suggestions/SuggestionCommand.kt
package main.commands.slashcommands.util.suggestions

import dev.minn.jda.ktx.coroutines.await
import dev.minn.jda.ktx.interactions.components.Modal
import dev.minn.jda.ktx.interactions.components.TextInput
import dev.minn.jda.ktx.interactions.components.danger
import dev.minn.jda.ktx.interactions.components.success
import dev.minn.jda.ktx.messages.EmbedBuilder
import dev.minn.jda.ktx.util.SLF4J
import main.utils.GeneralUtils.digits
import main.utils.GeneralUtils.dm
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.database.mongodb.cache.BotDBCache
import main.utils.locale.LocaleManager
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.SuggestionMessages
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.interactions.components.ActionRow
import net.dv8tion.jda.api.interactions.components.text.TextInputStyle
import java.awt.Color
import java.time.Instant

class SuggestionCommand : AbstractSlashCommand(
    SlashCommand(
        name = "suggest",
        description = "Suggest a feature you'd like to see in the bot.",
        guildUseOnly = false
    )
) {

    companion object {
        private val logger by SLF4J
    }

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        val localeManager = LocaleManager[event.guild]

        val textFieldPlaceholder = localeManager.getMessage(SuggestionMessages.SUGGESTION_MODAL_PLACEHOLDER)
        val suggestion = TextInput(
            id = "suggestion:suggestion_field",
            label = localeManager.getMessage(SuggestionMessages.SUGGESTION_MODAL_LABEL),
            placeholder = textFieldPlaceholder.substring(0, textFieldPlaceholder.length.coerceAtMost(100)),
            style = TextInputStyle.PARAGRAPH,
        )

        event.replyModal(
            Modal(
                id = "suggestion:modal",
                title = localeManager.getMessage(SuggestionMessages.SUGGESTION_MODAL_TITLE),
                components = listOf(ActionRow.of(suggestion))
            )
        ).queue()
    }

    override suspend fun onModalInteraction(event: ModalInteractionEvent) {
        if (event.modalId == "suggestion:modal") handleSuggestionModal(event)
        else if (event.modalId.startsWith("suggestion_reason:")) handleReasonModal(event)
    }

    private suspend fun handleSuggestionModal(event: ModalInteractionEvent) {
        val guild = event.guild
        val suggestion = event.getValue("suggestion:suggestion_field")?.asString
            ?: return event.replyEmbed(SuggestionMessages.INVALID_SUGGESTION).queue()

        val config = BotDBCache.instance
        val pendingChannel = event.jda.shardManager!!.getTextChannelById(config.suggestionsPendingChannelId)

        if (!config.suggestionSetup) {
            logger.warn("The suggestion channels aren't setup!")
            return event.replyEmbed(GeneralMessages.FEATURE_UNAVAILABLE)
                .setEphemeral(true)
                .queue()
        }

        if (config.userIsSuggestionBanned(event.user.idLong))
            return event.replyEmbed(SuggestionMessages.SUGGESTION_USER_BANNED)
                .setEphemeral(true)
                .queue()

        val suggester = event.user
        val embed = getGenericSuggestionEmbed(
            title = "Suggestion from ${suggester.name}",
            color = Color(255, 196, 0),
            suggester = suggester,
            suggestion = suggestion
        )

        pendingChannel!!.sendEmbed { embed }
            .setActionRow(
                success(
                    id = "suggestion_button:accept",
                    label = "Accept"
                ),
                danger(
                    id = "suggestion_button:deny",
                    label = "Deny"
                )
            )
            .queue()

        return event.replyEmbed(SuggestionMessages.SUGGESTION_SUBMITTED)
            .setEphemeral(true)
            .queue()
    }

    private suspend fun handleReasonModal(event: ModalInteractionEvent) {
        val config = BotDBCache.instance
        val guild = event.guild!!
        val localeManager = LocaleManager[guild]
        val (_, reasonType, messageId) = event.modalId.split(":")
        val reason = event.getValue("suggestion:reason_field")?.asString ?: if (reasonType == "accept")
            localeManager.getMessage(SuggestionMessages.DEFAULT_ACCEPT_REASON)
        else localeManager.getMessage(SuggestionMessages.DEFAULT_DENY_REASON)

        val shardManager = event.jda.shardManager!!
        val pendingChannel = shardManager.getTextChannelById(config.suggestionsPendingChannelId)!!
        event.deferReply(true).queue()

        val suggestionMessage = pendingChannel.retrieveMessageById(messageId).await()
        val suggester = shardManager.getUserById(suggestionMessage.embeds.first().fields.first().value!!.digits())
        val pendingSuggestion = suggestionMessage.embeds.first().description!!

        when (reasonType) {
            "accept" -> {
                val acceptedChannel = shardManager.getTextChannelById(config.suggestionsAcceptedChannelId)
                if (acceptedChannel == null) {
                    logger.warn("The accepted suggestion channel isn't setup!")
                    return event.replyEmbed(GeneralMessages.FEATURE_UNAVAILABLE)
                        .setEphemeral(true)
                        .queue()
                }

                acceptedChannel.sendMessageEmbeds(
                    getGenericSuggestionEmbed(
                        title = "Suggestion",
                        color = Color(77, 255, 69),
                        suggester = suggester,
                        suggestion = pendingSuggestion,
                        reason = reason
                    )
                ).queue()

                suggestionMessage.delete().queue()

                suggester?.dm(
                    EmbedBuilder(
                        title = "Suggestions",
                        color = Color(77, 255, 69).rgb,
                        description = "**Your suggestion has been accepted!**" +
                                "\nYou will see it appear in the next changelog\n\n**Your suggestion**\n```${
                                    pendingSuggestion.substring(
                                        0,
                                        pendingSuggestion.length.coerceAtMost(3900)
                                    )
                                }```",
                        fields = listOf(
                            MessageEmbed.Field("Comments", reason, false)
                        )
                    ).build()
                )

                event.hook.sendEmbed(guild, "Successfully accepted the suggestion!")
                    .queue()
            }

            "deny" -> {
                val deniedChannel = shardManager.getTextChannelById(config.suggestionsDeniedChannelId)
                if (deniedChannel == null) {
                    logger.warn("The denied suggestion channel isn't setup!")
                    return event.replyEmbed(GeneralMessages.FEATURE_UNAVAILABLE)
                        .setEphemeral(true)
                        .queue()
                }

                deniedChannel.sendMessageEmbeds(
                    getGenericSuggestionEmbed(
                        title = "Suggestion",
                        color = Color(187, 0, 0),
                        suggester = suggester,
                        suggestion = pendingSuggestion,
                        reason = reason
                    )
                ).queue()

                suggestionMessage.delete().queue()

                suggester?.dm(
                    EmbedBuilder(
                        title = "Suggestions",
                        color = Color(187, 0, 0).rgb,
                        description = "**Your suggestion has been rejected.**\n\n" +
                                "**Your suggestion**\n" +
                                "```${
                                    pendingSuggestion.substring(
                                        0,
                                        pendingSuggestion.length.coerceAtMost(3900)
                                    )
                                }```",
                        fields = listOf(
                            MessageEmbed.Field("Comments", reason, false)
                        )
                    ).build()
                )

                event.hook.sendEmbed(guild, "Successfully denied the suggestion!")
                    .queue()
            }
        }
    }

    override suspend fun onButtonInteraction(event: ButtonInteractionEvent) {
        if (!event.componentId.startsWith("suggestion_button:")) return

        val config = BotDBCache.instance
        val guild = event.guild
        if (!config.isDeveloper(event.user.idLong))
            return event.replyEmbed(GeneralMessages.NO_PERMS_BUTTON)
                .setEphemeral(true)
                .queue()

        val messageId = event.messageId
        val reasonModalText = TextInput(
            id = "suggestion:reason_field",
            label = "Reason",
            placeholder = "Fancy reason goes here",
            requiredLength = 0..1024,
            required = false,
            style = TextInputStyle.PARAGRAPH
        )

        val reasonModal = Modal(
            id = "suggestion_reason:${event.componentId.split(":")[1]}:$messageId",
            title = "Provide A Reason",
            components = listOf(ActionRow.of(reasonModalText))
        )

        event.replyModal(reasonModal).queue()
    }

    private fun getGenericSuggestionEmbed(
        title: String,
        color: Color,
        suggester: User?,
        suggestion: String,
        reason: String? = null
    ): MessageEmbed =
        EmbedBuilder(
            title = title,
            fields = listOf(
                MessageEmbed.Field("Suggester", suggester?.asMention ?: "Unknown Suggester", false),
            ) + if (reason != null) listOf(
                MessageEmbed.Field("Reason", reason, false)
            ) else listOf(),
            description = suggestion,
            color = color.rgb,
            timestamp = Instant.now()
        ).build()

}